### PR TITLE
💸 Token transfer amount handling

### PR DESCRIPTION
**Pull request by jodator**
In this PR:

- user can select arbitrary value in token transfer modal
  - set half/max balance value
- few UI components were refactored as some were duplicated

However, I still face the problem with the development setup as I wan't able to set working environment other than default substrate build on my Apple M1 :/

### CODE REVIEW

File: packages/ui/test/hooks/useNumberInput.test.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import { expect } from 'chai'
+import { renderHook, act } from '@testing-library/react-hooks'
+import { useNumberInput } from '../../src/hooks/useNumberInput'
+
+describe('useNumberInput', () => {
+  function render() {
+    let decimals = 18
+    const result = renderHook(() => useNumberInput(decimals))
```
Comment (krzysztof-jelski) [inline packages/ui/test/hooks/useNumberInput.test.ts:8]: Propagate to destructuring, should look cleaner.

File: packages/ui/src/utils/formatters.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import BN from 'bn.js'
+import { formatBalance } from '@polkadot/util'
+
+const decimals = 12
```
Comment (krzysztof-jelski) [inline packages/ui/src/utils/formatters.ts:4]:
```suggestion
const DECIMALS = 12
```
(aka convention to have constants all caps)

File: packages/ui/src/hooks/useNumberInput.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  }, [decimals])
+
+  return [value, set] as const
+}
+
+function stripLeadingZeroes(value: string) {
+  if (value === '') {
+    return value
+  }
+  const leadingZeroes = value.match(/^0*/)?.[0].length
```
Comment (krzysztof-jelski) [inline packages/ui/src/hooks/useNumberInput.ts:25]:
```suggestion
  const leadingZeroes = value.search(/^0*/) >= 0
```

File: packages/ui/src/hooks/useNumberInput.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  if (value.startsWith('.') || value === '') {
+    value = '0' + value
+  }
+  return value
+}
+
+function truncateDecimals(value: string, decimals: number) {
+  if (value.includes('.')) {
+    const [integer, decimal] = value.split('.')
+    const trimmed = decimal.substring(0, decimals)
```
Comment (krzysztof-jelski) [inline packages/ui/src/hooks/useNumberInput.ts:38]: Shouldn't we have some rounding here? Eg. trim(1.29, 1) -> 1.3

File: packages/ui/src/components/AccountInfo.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+//   width: fit-content;
+//   margin: 0;
+//   padding: 0 8px;
+//   font-size: 10px;
+//   line-height: 16px;
+//   border-radius: 8px;
+//   color: ${Colors.White};
+//   background-color: ${Colors.Blue[200]};
+//   text-transform: uppercase;
+// `
```
Comment (krzysztof-jelski) [inline packages/ui/src/components/AccountInfo.tsx:89]: Remove, I guess

File: packages/ui/src/components/TransferModal.tsx [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+export function TransferModal({ from, to, onClose }: Props) {
+  const { api } = useApi()
+  const { keyring } = useKeyring()
+  const balances = useBalances([from, to])
+  const [isSending, setIsSending] = useState(false)
+  const [amount, setAmount] = useNumberInput(0)
+
+  if (!balances.hasBalances)
```
Comment (krzysztof-jelski) [inline packages/ui/src/components/TransferModal.tsx:28]: I suggest adding braces.

File: packages/ui/src/components/TransferModal.tsx [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+  const setHalf = () => setAmount(transferableBalance.div(new BN(2)).toString())
+  const setMax = () => setAmount(transferableBalance.toString())
+  const signAndSend = async () => {
+    setIsSending(true)
+
+    if (!api || !keyring) {
+      return
+    }
+
+    await api.tx.balances
```
Comment (krzysztof-jelski) [inline packages/ui/src/components/TransferModal.tsx:50]: Consider adding a UI test for this modal. I'm not insisting on it, but could trigger some healthy refactoring (eg. not depending on api directly).

**Review by krzysztof-jelski — APPROVED** [review]

File: packages/ui/src/hooks/useNumberInput.ts (hunk 4 above)
Comment (jodator) [reply packages/ui/src/hooks/useNumberInput.ts:38]: I've based this on dapp-boilerplate. The number imput is used to validate user input, so rounding might not be desired here.

File: packages/ui/src/components/TransferModal.tsx (hunk 7 above)
Comment (jodator) [reply packages/ui/src/components/TransferModal.tsx:51]: That makes sense. I'll add them in the next task.

**Pull request merged**